### DISCUSSION TITLE
How do I actually install Neolink after Gstreamer is installed on Ubuntu Server?

### DISCUSSION THREAD

**Discussion by aaronj1234** [thirtythreeforty/neolink#126] [Q&A]
Hello,
I don't really understand how to install the Neolink package with the configuration file that I made. It just says to open a shell and type in that command "neolink --config my_config.toml" but it just comes up with an error, even if I cd to the directory where I unzipped it.

**Comment by QuantumEntangledAndy** [comment]
Oh my apologies on ubuntu, could you try this

1. cd to the directory it was unzipped to
2. make it exceutable
```bash
chmod +x neolink
```
3. Copy the my_config.toml to same directory
4. Run it with this
```bash
./neolink --config my_config.toml
```

Note the `./` at the front

**Reply by QuantumEntangledAndy** [reply]
If you get any errors please paste what it says here

**Reply by aaronj1234** [reply]
Hello, 
I tried the above commands, but I just got the following errors:
![image](https://user-images.githubusercontent.com/77698398/105145479-b597a100-5af6-11eb-94de-004efbb2bd20.png)
(The extra commands were my attempts)
I also tried installing Gstreamer again, but there was nothing else to install

**Reply by aaronj1234** [reply]
I also tried becoming super user with 'sudo su' but it was still the same issue

**Reply by QuantumEntangledAndy** [reply]
Permission denied suggests that the chmod failed.

Can you run

```bash
ls -la .
```

in the folder with neolink.

I am suspecting that executable bit is not setting up correctly perhaps because of some other secutiry constraint

**Reply by QuantumEntangledAndy** [reply]
I suspect that the download dir has been setup with the `noexec mount` option. Probably as a security feature. If you could run that `ls -la` above and show us the result and maybe also show the output of `mount` that will tell us

**Reply by aaronj1234** [reply]
Hello,
Here are the outputs for those commands.
ls -la:
![image](https://user-images.githubusercontent.com/77698398/105148545-6c495080-5afa-11eb-827e-2e38450ab1fa.png)

mount:
![image](https://user-images.githubusercontent.com/77698398/105148632-85520180-5afa-11eb-9616-56f69d18f5ef.png)

**Reply by QuantumEntangledAndy** [reply]
Do you see the part in the `ls -la` where it says `-rw-rw-r--` this part means:
- 1st char special usually the sticky bit or the suid bit almost always `-`
- 2nd char `r` means owner can read
- 3rd char `w` means owner can write
- 4th char `-` means owner CANNOT exec <-- this is our error it should say `x`
- 5th char `r` means group can read
- 6th char `w` means group can write
- 7th char `-` means group cannot exec
- 8th char `r` means anyone else can read
- 9th char `-` means noone else can write (this is good)
- 10th char `-` means noone else can exec

You cannot run it because the chmod +x was not able to add the `x` to the 4th char. I suspect this is a security feature. Can you try and move it out of downloads into any other folder. Maybe

```bash
sudo cp ~/Downloads/neolink /usr/local/bin/neolink
sudo chmod +x /usr/local/bin/neolink
ls -la /usr/local/bin/neolink
```

then show me the results

**Reply by aaronj1234** [reply]
Hello,
Here's the results:
![image](https://user-images.githubusercontent.com/77698398/105157860-4d03f080-5b05-11eb-9e88-e9fd31a908df.png)

The results look better than the previous results as there is no `-` as the 4th char, just an `x` so I'm guessing I can now make it executable

**Reply by QuantumEntangledAndy** [reply]
Yep that is exec. I think your download dir has the noexec security on it. 

We copied it to `/usr/local/bin` because that is a special directory (it is part of $PATH by default). You can run programs in there from any folder without using the `./` prefix.

So just run

```bash
neolink --config my_config.toml
```

No need for `./` or even for neolink to be in the same folder because ubuntu will use the one from the `/usr/local/bin`

**Reply by aaronj1234** [reply]
Hello,
I got another few errors :(
![image](https://user-images.githubusercontent.com/77698398/105168961-c2c28900-5b12-11eb-9614-44aa777fc18b.png)
The configuration file is correct, so I'm not really sure what's going on

**Reply by QuantumEntangledAndy** [reply]
I suspect there is something wrong with the config. Can you show it to us? just `***` out the sensitive stuff. I suspect you have missed a closing `"` or something small

**Reply by aaronj1234** [reply]
Hello,
Here's my config file:
![image](https://user-images.githubusercontent.com/77698398/105173394-da047500-5b18-11eb-9da4-a92818d3dd42.png)

**Reply by QuantumEntangledAndy** [reply]
That looks ok but lets go back a step. Why did you copy 

`/Neolink`

This is not the executable I think. I think this is something different.

The excutable should be about `15499080` bytes but the one in your `/usr/local/bin/neolink` is only `161` bytes

**Comment by QuantumEntangledAndy** [comment]
Please can you run this:

```bash
ls -lha ~/Downloads/neolink /usr/local/bin/neolink
ls -lha /usr/local/bin/neolink
```

They should have the same file size

**Reply by aaronj1234** [reply]
Hello again,
I did the above commands, and so I copied the `neolink.d` file to the `/usr/local/bin/neolink` directory, and I only got 1 error:
![image](https://user-images.githubusercontent.com/77698398/105179018-b0e7e280-5b20-11eb-8b80-723622e837f1.png)

**Reply by QuantumEntangledAndy** [reply]
Do not copy neolink.d but instead copy neolink.

In the download directory run

```
sudo cp neolink /usr/local/bin/neolink
```

**Reply by QuantumEntangledAndy** [reply]
The reason it did not work is because earlier I asked you to do this:

> ```shell
> sudo cp ~/Downloads/neolink /usr/local/bin/neolink # <- Right command
> sudo chmod +x /usr/local/bin/neolink
> ls -la /usr/local/bin/neolink
> ```

But instead you incorrectly did this:

```bash
sudo cp /Neolink /usr/local/bin/neolink  # <- Wrong command
```

**Reply by aaronj1234** [reply]
Hi again,
I did the correct commands, and I got a different error:
![image](https://user-images.githubusercontent.com/77698398/105186013-248ded80-5b29-11eb-9271-f893fc2e863f.png)

**Reply by QuantumEntangledAndy** [reply]
You do not have gstreamer installed. Please install it with this

```bash
sudo apt install \
  libgstrtspserver-1.0-0 \
  libgstreamer1.0-0 \
  libgstreamer-plugins-bad1.0-0 \
  gstreamer1.0-plugins-good \
  gstreamer1.0-plugins-bad
```

**Reply by aaronj1234** [reply]
Hello again,
I got a brand new error:
![image](https://user-images.githubusercontent.com/77698398/105188487-ddedc280-5b2b-11eb-96f2-d8501e810108.png)

I installed gstreamer, but now there's that new error.
And also, once it's fixed, because the camera's called Path Camera, when I try to connect to it, do I connect to the camera's IP in the rtsp address or the server's address?

**Reply by QuantumEntangledAndy** [reply]
Your current error means it cannot find the file `my_config.toml` make sure it is called that and in the right place.

The IP if the neolink server. 

You should try to connect to the camera first using local host from the Ubuntu machine.

```bash
ffprobe rtsp://127.0.0.1/camera_name
```

`127.0.0.1` is a special ip that means this machine. So it will only work from the same machine neolink is installed. We test first using this to make sure the firewall or other neowork issues aregetting in the way.

**Reply by QuantumEntangledAndy** [reply]
It's getting late where I am going to sleep goodlucj and hopefully you can get it fixed.

**Reply by aaronj1234** [reply]
Thanks for your help Andy, and goodnight

**Reply by QuantumEntangledAndy** [reply]
How did it go any success?

**Reply by aaronj1234** [reply]
Hello,
Unfortunately, I've got a couple more errors:
![image](https://user-images.githubusercontent.com/77698398/105341066-94a67d00-5bd6-11eb-8ef3-473f0698fc3a.png)

**Reply by aaronj1234** [reply]
I also opened the `neolink` file in the `usr/local/bin` dir, and it looked normal:
![image](https://user-images.githubusercontent.com/77698398/105342284-121ebd00-5bd8-11eb-8cc9-4e15f257f83f.png)

**Comment by QuantumEntangledAndy** [comment]
You have gone back a step ToT. You have copied the neolink CONFIG into `/usr/local/bin/neolink`. The file at `/usr/local/bin/neolink` SHOULD BE THE BINARY. We fixed this yesterday and you have unfixed it since then! Please do this:

```bash
cd ~/Downloads/
cp /usr/local/bin/neolink my_config.toml
sudo cp neolink /usr/local/bin/neolink
sudo chmod +x /usr/local/bin/neolink
neolink --config my_config.toml
```

**Reply by QuantumEntangledAndy** [reply]
Confirm that this file at `/usr/local/bin/neolink` is the right one by showing me the output of:

```bash
ls -lah /usr/local/bin/neolink
ls -lah my_config.toml
```

**Reply by aaronj1234** [reply]
Good news: It's working!
![image](https://user-images.githubusercontent.com/77698398/105345341-0df49e80-5bdc-11eb-816f-8060b30becd9.png)
The camera's internet connection is a little intermittent at times, but at least the command is working!
Thanks for your help :)

**Reply by QuantumEntangledAndy** [reply]
Oh good. Since you screen shoted your password please change it. Also if you want neolink to keep running in the background please consider reading these documents:

https://github.com/thirtythreeforty/neolink/blob/master/docs/unix_setup.md

https://github.com/thirtythreeforty/neolink/blob/master/docs/unix_service.md

**Reply by aaronj1234** [reply]
Oh gosh did I? I'll change it immediately :l

**Reply by aaronj1234** [reply]
Thanks!
